beacon-chain/sync/subscription_topic_handler.go

```diff
@@ -0,0 +1,90 @@
+package sync
+
+import (
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+)
+
+// This is a subscription topic handler that is used to handle basic
+// CRUD operations on the topic map. All operations are thread safe
+// so they can be called from multiple routines.
+type subTopicHandler struct {
+	sync.RWMutex
+	subTopics map[string]*pubsub.Subscription
+	digestMap map[[4]byte]int
+}
+
+func newSubTopicHandler() *subTopicHandler {
+	return &subTopicHandler{
+		subTopics: map[string]*pubsub.Subscription{},
+		digestMap: map[[4]byte]int{},
+	}
+}
+
+func (s *subTopicHandler) addTopic(topic string, sub *pubsub.Subscription) {
+	s.Lock()
+	defer s.Unlock()
+	s.subTopics[topic] = sub
+	digest, err := p2p.ExtractGossipDigest(topic)
+	if err != nil {
```

Comment:
Potential race condition. if this errs out, we will be stuck with `s.subTopics[topic] == sub` but no update to `digestMap`

Reply:
I see your following comment @jmozah. Sounds like it's intentional, even `digestMap` fails we should still update the subscription map